src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift

```diff
@@ -25,53 +25,104 @@ final class ExposureDetection {
 	private var progress: Progress?
 
+	private var countryKeypackageDownloader: CountryKeypackageDownloading
+
 	// MARK: Creating a Transaction
-	init(delegate: ExposureDetectionDelegate) {
+	init(
+		delegate: ExposureDetectionDelegate,
+		countryKeypackageDownloader: CountryKeypackageDownloading? = nil
+	) {
 		self.delegate = delegate
+
+		if let countryKeypackageDownloader = countryKeypackageDownloader {
+			self.countryKeypackageDownloader = countryKeypackageDownloader
+		} else {
+			self.countryKeypackageDownloader = CountryKeypackageDownloader(delegate: delegate)
+		}
 	}
 
 	func cancel() {
 		progress?.cancel()
 	}
 
-	// MARK: Starting the Transaction
-	// Called right after the transaction knows which data is available remotly.
-	private func downloadDeltaUsingAvailableRemoteData(_ remote: DaysAndHours?) {
-		guard let remote = remote else {
-			endPrematurely(reason: .noDaysAndHours)
-			return
-		}
-		guard let delta = delegate?.exposureDetection(self, downloadDeltaFor: remote) else {
-			endPrematurely(reason: .noDaysAndHours)
-			return
-		}
-		delegate?.exposureDetection(self, downloadAndStore: delta) { [weak self] error in
+	private func getSupportedCountries(completion: @escaping ([Country]) -> Void) {
+		delegate?.exposureDetection(supportedCountries: { [weak self] result in
 			guard let self = self else { return }
-			if error != nil {
-				self.endPrematurely(reason: .noDaysAndHours)
-				return
+
+			switch result {
+			case .success(let supportedCountries):
+				completion(supportedCountries)
+			case.failure:
+				self.endPrematurely(reason: .noSupportedCountries)
 			}
-			self.delegate?.exposureDetection(self, downloadConfiguration: self.useConfiguration)
-		}
+		})
+	}
+
+	private func getCountriesToDetect(supportedCountries: [Country]) -> [Country.ID] {
+		var countryIDs = supportedCountries.map { $0.id }
+		countryIDs.append(Country.defaultCountry().id)
+		return countryIDs
 	}
 
-	private func useConfiguration(_ configuration: ENExposureConfiguration?) {
-		guard let configuration = configuration else {
-			endPrematurely(reason: .noExposureConfiguration)
-			return
+	private func downloadKeyPackages(for countries: [Country.ID], completion: @escaping () -> Void) {
+		let dispatchGroup = DispatchGroup()
+		var errors = [ExposureDetection.DidEndPrematurelyReason]()
+
+		for country in countries {
+			dispatchGroup.enter()
+
+			self.countryKeypackageDownloader.downloadKeypackages(for: country) { result in
+				switch result {
+				case .failure(let didEndPrematurelyReason):
+					errors.append(didEndPrematurelyReason)
+				case .success:
+					break
+				}
+
+				dispatchGroup.leave()
+			}
 		}
-		guard let writtenPackages = delegate?.exposureDetectionWriteDownloadedPackages(self) else {
-			endPrematurely(reason: .unableToWriteDiagnosisKeys)
-			return
+
+		dispatchGroup.notify(queue: .main) {
+			if let error = errors.first {
+				self.endPrematurely(reason: error)
+			} else {
+				completion()
+			}
```

Comment:
why do we collect all errors and only return the first one?

Reply:
Its a dispatch group and the UI can only handle one error. If one error occurs, risk calculation will not be triggered.

Reply:
I have added 2 tests for this unhappy path.